core/schema/env.graphqls

```diff
@@ -0,0 +1,252 @@
+  checks: [Check!]
+
+  "The check in this environment with the given name, if any"
+  check(name: String!): Check
+}
+
+"An entrypoint for tests, lints or anything that can pass/fail."
+type Check {
+  "A unique identifier for this check."
+  id: CheckID!
+
+  "The name of the check."
+  name: String!
+
+  "This check with the given name"
+  withName(name: String!): Check!
+
+  "Documentation for this check."
+  description: String
+
+  "This check with the given description"
+  withDescription(description: String!): Check!
+
+  "The subchecks of this check."
+  subchecks: [Check!]
+
+  "This check with the given subcheck"
+  withSubcheck(id: CheckID!): Check!
+
+  """
+  This check with the given container used to determine the check's result.
+  If set, the container will be executed and the check success will be true
+  if the container exits with a zero exit code, false otherwise.
+  """
+  withContainer(id: ContainerID!): Check!
+
+  "The result of evaluating this check."
+  result: CheckResult!
+}
+
+type CheckResult {
+  "A unique identifier for this check result."
+  id: CheckResultID!
+
+  "Whether the check was successful."
+  success: Boolean!
+
+  "Any output obtained from evaluating the check's success."
+  output: String!
+}
+
+extend type Environment {
+  """
+  Internal only.
+
+  This environment with the given check. It is an error to call this outside
+  of environment initialization code.
+  """
+  withCheck(id: CheckID!, returnType: CheckEntrypointReturnType): Environment!
+
+  """
+  Internal only.
+
+  The input for the current entrypoint invocation.
+  """
+  entrypointInput: EntrypointInput!
+
+  """
+  Internal only.
+
+  Return the given value as the result of the current entrypoint invocation.
+  The value is expected to be the json-serialized representation of the return.
+  """
+  returnEntrypointValue(value: String!): Boolean!
+}
+
+"""
+Internal-only.
+
+The input provided to an SDK entrypoint invocation.
+"""
+type EntrypointInput {
+  """
+  Internal-only.
+
+  The name of the entrypoint to invoke. If not set, then the sdk
+  is expected to return the environment definition.
+  """
+  name: String
+
+  """
+  Internal-only.
+
+  The json-serialized arguments to pass to the entrypoint. The json
+  object is a map of argument names to argument values.
+  """
+  args: String!
+}
+
+"""
+Internal-only.
+
+The handled return types for a check entrypoint.
+"""
+enum CheckEntrypointReturnType {
+  """
+  Internal-only.
+
+  The entrypoint does not return any value.
+
+  In this case, the check's result is based on whether the entrypoint exit code.
+
+  Exit code 0 will result in a successful check.
+
+  Exiting with exit code 1 will result in a failed check, but the entrypoint
+  execution will be cached.
+
+  Exiting with any other exit code will be treated an internal error and not
+  result in the entrypoint execution being cached. This error will be forwarded
+  to any clients querying the check result.
```

Comment:
Checking my understanding: we're able to be really restrictive here because the other side will always be implemented by one of our runtimes who uphold their end of this contract, rather than an exit code produced by running tests.

Reply:
Yep exactly, for a little more background, the constraints we're working with here are:
1. We want SDKs to support multiple return types for a given entrypoint 
2. We want SDKs to be as dumb as possible and implement as much server-side as possible
3. For whatever SDKs do need to do, it's best if it's encoded in the api; or at min doc'd as part of that api
4. For checks, we need to distinguish between "internal errors" (which should not result in execution being cached) and "check errors" (which should result in the error being cached)

So the schema you commented on here is basically where I landed in terms of the engine<->sdk contract. SDKs need to tell the server what the specific entrypoint is returning (`Check`, `CheckResult`, `string`, void) and use different exit codes to distinguish between "internal error" and "user code error".
* I had a few different approaches to this earlier, but of the iterations that actually worked, this one resulted in the simplest code for the Go SDK

After that, the server does the rest of the magic needed to get the caching right. As part of this I actually added an internal ability for `Container.WithExec` to treat cache execution of execs with non-zero exit codes: https://github.com/sipsma/dagger/blob/179525ff35a1e4c5b338472e47d350d4dbee8fb2/core/container.go#L1855-L1855

Which could hopefully come in use later if we want to re-use that as part of our public api.